File: app/src/main/java/com/majeur/psclient/model/BattleDecision.java
package com.majeur.psclient.model;

import java.util.LinkedList;
import java.util.List;

import static com.majeur.psclient.util.Utils.str;

public class BattleDecision {

    private static final String CMD_CHOOSE = "choose";
    private static final String CMD_TEAM = "team";
    private static final String ACTION_MOVE = "move";
    private static final String ACTION_SWITCH = "switch";
    private static final String ACTION_PASS = "pass";
    private static final String EXTRA_MEGA = "mega";
    private static final String EXTRA_ZMOVE = "zmove";
    private static final String EXTRA_DYNAMAX = "dynamax";

    private String mCommand;
    private final List<Choice> mChoices = new LinkedList<>();

    private static class Choice {
        String action;
        int index;
        String extra;
        int target;
    }

    public void addSwitchChoice(int who) {
        mCommand = CMD_CHOOSE;
        Choice c = new Choice();
        c.action = ACTION_SWITCH;
        c.index = who;
        mChoices.add(c);
    }

    public void addMoveChoice(int which, boolean mega, boolean zmove, boolean dynamax) {
        mCommand = CMD_CHOOSE;
        Choice c = new Choice();
        c.action = ACTION_MOVE;
        c.index = which;
        if (mega) c.extra = EXTRA_MEGA;
        else if (zmove) c.extra = EXTRA_ZMOVE;
        else if (dynamax) c.extra = EXTRA_DYNAMAX;
        mChoices.add(c);
    }

    public void setLastMoveTarget(int target) {
        mChoices.get(mChoices.size() - 1).target = target;
    }

    public void addPassChoice() {
        mCommand = CMD_CHOOSE;
        Choice c = new Choice();
        c.action = ACTION_PASS;
        mChoices.add(c);
    }

    public void addLeadChoice(int first, int teamSize) {
        mCommand = CMD_TEAM;
        Choice c = new Choice();
        c.index = first;
        c.target = teamSize;
        mChoices.add(c);
    }

    public int leadChoicesCount() {
        int count = 0;
        for (Choice c : mChoices)
            if (c.action == null)
                count++;
        return count;
    }

    public int switchChoicesCount() {
        int count = 0;
        for (Choice c : mChoices)
            if (c.action.equals(ACTION_SWITCH))
                count++;
        return count;
    }

    public boolean hasSwitchChoice(int which) {
        for (Choice c : mChoices)
            if (c.action.equals(ACTION_SWITCH) && c.index == which)
                return true;
        return false;
    }

    public boolean hasOnlyPassChoice() {
        for (Choice c : mChoices)
            if (!c.action.equals(ACTION_PASS))
                return false;
        return true;
    }

    public String getCommand() {
        return mCommand;
    }

    public String build() {
        if (mCommand.equals(CMD_TEAM)) return buildLeadString();
        StringBuilder builder = new StringBuilder();
        int N = mChoices.size();
        for (int i = 0; i < N; i++) {
            Choice c = mChoices.get(i);
            builder.append(c.action);
            if (c.index != 0) builder.append(" ").append(c.index);
            if (c.extra != null) builder.append(" ").append(c.extra);
            if (c.target != 0) builder.append(" ").append(c.target);
            if (i < N -1) builder.append(",");
        }
        return builder.toString();
    }

    private String buildLeadString() {
        StringBuilder builder = new StringBuilder();
        int teamSize = 0;
        for (Choice c : mChoices) {
            builder.append(c.index);
            teamSize = c.target;
        }
        for (int i = 1; i <= teamSize; i++)
            if (builder.indexOf(str(i)) < 0) builder.append(i);
        return builder.toString();
    }
}

File: psclient/src/main/java/com/majeur/psclient/model/battle/BattleDecision.kt
package com.majeur.psclient.model.battle


class BattleDecision {

    private var _command: String? = null

    val command: String get() {
        if (_command == null) throw NullPointerException("BattleDecision has no command set")
        return _command!!
    }

    private val choices = mutableListOf<Choice>()
    private var teamSize = 0

    private data class Choice(
            val action: String = "",
            val index: Int = 0,
            val extra: String? = null,
            var target: Int = 0
    )

    /* 1 based index */
    fun addSwitchChoice(who: Int) {
        _command = CMD_CHOOSE
        choices.add(Choice(action = ACTION_SWITCH, index = who))
    }

    /* 1 based index */
    fun addMoveChoice(which: Int, mega: Boolean, zmove: Boolean, dynamax: Boolean) {
        _command = CMD_CHOOSE
        val extra = if (mega) EXTRA_MEGA else if (zmove) EXTRA_ZMOVE else if (dynamax) EXTRA_DYNAMAX else null
        choices.add(Choice(action = ACTION_MOVE, index = which, extra = extra))
    }

    /* 1 based index */
    fun setLastMoveTarget(target: Int) {
        choices.last().target = target
    }

    fun addPassChoice() {
        _command = CMD_CHOOSE
        choices.add(Choice(action = ACTION_PASS))
    }

    /* 1 based index */
    fun addLeadChoice(first: Int, teamSize: Int) {
        _command = CMD_TEAM
        this.teamSize = teamSize
        choices.add(Choice(index = first))
    }

    fun leadChoicesCount() = choices.count { it.action.isEmpty() }

    fun switchChoicesCount() = choices.count { it.action == ACTION_SWITCH }

    /* 1 based index */
    fun hasSwitchChoice(which: Int) = choices.any { it.action == ACTION_SWITCH && it.index == which }

    fun hasOnlyPassChoice() = choices.all { it.action == ACTION_PASS }

    fun build(): String {
        return StringBuilder().run {
            if (_command == CMD_TEAM) {
                choices.forEach { c -> append(c.index) }
                (1..teamSize).forEach { if (!contains(it.toString())) append(it) }
            } else {
                choices.forEach { c ->
                    append(c.action)
                    if (c.index != 0) append(" ").append(c.index)
                    if (c.extra != null) append(" ").append(c.extra)
                    if (c.target != 0) append(" ").append(c.target)
                    append(",")
                }
            }
            toString().removeSuffix(",")
        }
    }

    companion object {
        private const val CMD_CHOOSE = "choose"
        private const val CMD_TEAM = "team"
        private const val ACTION_MOVE = "move"
        private const val ACTION_SWITCH = "switch"
        private const val ACTION_PASS = "pass"
        private const val EXTRA_MEGA = "mega"
        private const val EXTRA_ZMOVE = "zmove"
        private const val EXTRA_DYNAMAX = "dynamax"
    }
}